From e4c7d2c06a339edc6ac59afe3a9e2e6b3751c027 Mon Sep 17 00:00:00 2001
From: Marvin Schmidt <marv@exherbo.org>
Date: Sun, 4 Nov 2018 18:35:57 +0100
Subject: [PATCH 1/2] pam_exec: Replace glibc specific strndupa usage
Upstream: Issue: https://github.com/linux-pam/linux-pam/issues/32
Upstream: PR: https://github.com/linux-pam/linux-pam/pull/66 (different patch)

Based on a patch from Void Linux by Juan RP <xtraeme@gmail.com>
---
 modules/pam_exec/pam_exec.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/modules/pam_exec/pam_exec.c b/modules/pam_exec/pam_exec.c
index 52dc681..bb5070e 100644
--- a/modules/pam_exec/pam_exec.c
+++ b/modules/pam_exec/pam_exec.c
@@ -103,11 +103,14 @@ call_exec (const char *pam_type, pam_handle_t *pamh,
   int optargc;
   const char *logfile = NULL;
   const char *authtok = NULL;
+  char authtok_buf[PAM_MAX_RESP_SIZE+1];
+
   pid_t pid;
   int fds[2];
   int stdout_fds[2];
   FILE *stdout_file = NULL;
 
+  memset (authtok_buf, 0, sizeof(authtok_buf));
   if (argc < 1) {
     pam_syslog (pamh, LOG_ERR,
 		"This module needs at least one argument");
@@ -180,12 +183,12 @@ call_exec (const char *pam_type, pam_handle_t *pamh,
 	      if (resp)
 		{
 		  pam_set_item (pamh, PAM_AUTHTOK, resp);
-		  authtok = strndupa (resp, PAM_MAX_RESP_SIZE);
+		  authtok = strncpy (authtok_buf, resp, sizeof(authtok_buf));
 		  _pam_drop (resp);
 		}
 	    }
 	  else
-	    authtok = strndupa (void_pass, PAM_MAX_RESP_SIZE);
+	    authtok = strncpy (authtok_buf, void_pass, sizeof(authtok_buf));
 
 	  if (pipe(fds) != 0)
 	    {
-- 
2.19.1

